File: aeron-driver/src/main/java/uk/co/real_logic/aeron/driver/buffer/MappedRawLog.java
/*
 * Copyright 2014 - 2015 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.aeron.driver.buffer;

import uk.co.real_logic.aeron.driver.event.EventCode;
import uk.co.real_logic.aeron.driver.event.EventLogger;
import uk.co.real_logic.aeron.logbuffer.LogBufferPartition;
import uk.co.real_logic.agrona.IoUtil;
import uk.co.real_logic.agrona.concurrent.UnsafeBuffer;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.stream.Stream;

import static java.nio.channels.FileChannel.MapMode.READ_WRITE;
import static uk.co.real_logic.aeron.logbuffer.LogBufferDescriptor.*;

/**
 * Encapsulates responsibility for mapping the files into memory used by the log partitions.
 */
class MappedRawLog implements RawLog
{
    private static final int ONE_GIG = 1 << 30;

    private final int termLength;
    private final LogBufferPartition[] partitions;
    private final EventLogger logger;
    private final File logFile;
    private final MappedByteBuffer[] mappedBuffers;
    private final UnsafeBuffer logMetaDataBuffer;

    MappedRawLog(final File location, final FileChannel blankTemplate, final int termLength, final EventLogger logger)
    {
        this.termLength = termLength;
        this.logger = logger;
        this.logFile = location;
        partitions = new LogBufferPartition[PARTITION_COUNT];

        try (final RandomAccessFile raf = new RandomAccessFile(logFile, "rw");
             final FileChannel logChannel = raf.getChannel())
        {
            final long logLength = computeLogLength(termLength);
            raf.setLength(logLength);
            blankTemplate.transferTo(0, logLength, logChannel);

            if (logLength <= Integer.MAX_VALUE)
            {
                final MappedByteBuffer mappedBuffer = logChannel.map(READ_WRITE, 0, logLength);
                mappedBuffers = new MappedByteBuffer[]{ mappedBuffer };
                final int metaDataSectionOffset = termLength * PARTITION_COUNT;

                for (int i = 0; i < PARTITION_COUNT; i++)
                {
                    final int metaDataOffset = metaDataSectionOffset + (i * TERM_META_DATA_LENGTH);

                    partitions[i] = new LogBufferPartition(
                        new UnsafeBuffer(mappedBuffer, i * termLength, termLength),
                        new UnsafeBuffer(mappedBuffer, metaDataOffset, TERM_META_DATA_LENGTH));
                }

                logMetaDataBuffer = new UnsafeBuffer(mappedBuffer, (int)(logLength - LOG_META_DATA_LENGTH), LOG_META_DATA_LENGTH);
            }
            else
            {
                mappedBuffers = new MappedByteBuffer[PARTITION_COUNT + 1];
                final long metaDataSectionOffset = termLength * (long)PARTITION_COUNT;
                final int metaDataSectionLength = (int)(logLength - metaDataSectionOffset);

                final MappedByteBuffer metaDataMappedBuffer = logChannel.map(
                    READ_WRITE, metaDataSectionOffset, metaDataSectionLength);
                mappedBuffers[mappedBuffers.length - 1] = metaDataMappedBuffer;

                for (int i = 0; i < PARTITION_COUNT; i++)
                {
                    mappedBuffers[i] = logChannel.map(READ_WRITE, termLength * (long)i, termLength);

                    partitions[i] = new LogBufferPartition(
                        new UnsafeBuffer(mappedBuffers[i]),
                        new UnsafeBuffer(metaDataMappedBuffer, i * TERM_META_DATA_LENGTH, TERM_META_DATA_LENGTH));
                }

                logMetaDataBuffer = new UnsafeBuffer(
                    metaDataMappedBuffer, metaDataSectionLength - LOG_META_DATA_LENGTH, LOG_META_DATA_LENGTH);
            }
        }
        catch (final IOException ex)
        {
            throw new IllegalStateException(ex);
        }
    }

    public int termLength()
    {
        return termLength;
    }

    public void close()
    {
        for (final MappedByteBuffer buffer : mappedBuffers)
        {
            IoUtil.unmap(buffer);
        }

        if (!logFile.delete())
        {
            logger.log(EventCode.ERROR_DELETING_FILE, logFile);
        }
    }

    public Stream<LogBufferPartition> stream()
    {
        return Stream.of(partitions);
    }

    public LogBufferPartition[] partitions()
    {
        return partitions;
    }

    public UnsafeBuffer logMetaData()
    {
        return logMetaDataBuffer;
    }

    public ByteBuffer[] sliceTerms()
    {
        final ByteBuffer[] terms = new ByteBuffer[PARTITION_COUNT];
        final int termLength = partitions[0].termBuffer().capacity();

        if (termLength < ONE_GIG)
        {
            final MappedByteBuffer buffer = mappedBuffers[0];
            for (int i = 0; i < PARTITION_COUNT; i++)
            {
                buffer.limit((termLength * i) + termLength).position(termLength * i);
                terms[i] = buffer.slice();
            }
        }
        else
        {
            for (int i = 0; i < PARTITION_COUNT; i++)
            {
                terms[i] = mappedBuffers[i].duplicate();
            }
        }

        return terms;
    }

    public String logFileName()
    {
        return logFile.getAbsolutePath();
    }
}

File: aeron-driver/src/main/cpp/buffer/MappedRawLog.cpp
/*
 * Copyright 2016 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <unistd.h>
#include "MappedRawLog.h"

namespace aeron { namespace driver { namespace buffer {

MappedRawLog::MappedRawLog(
    const char* location,
    bool useSparseFiles,
    std::int32_t termLength) :
    m_location{location},
    m_useSparseFiles{useSparseFiles},
    m_termLength{termLength}
{
    std::int64_t logLength = LogBufferDescriptor::computeLogLength(termLength);

    if (logLength < LogBufferDescriptor::MAX_SINGLE_MAPPING_SIZE)
    {
        m_memoryMappedFiles.push_back(MemoryMappedFile::createNew(m_location, 0, (size_t) logLength));

        std::uint8_t *basePtr = m_memoryMappedFiles[0]->getMemoryPtr();

        if (!useSparseFiles)
        {
            allocatePages(basePtr, m_memoryMappedFiles[0]->getMemorySize());
        }

        for (int i = 0; i < LogBufferDescriptor::PARTITION_COUNT; i++)
        {
            m_buffers[i].wrap(basePtr + (i * termLength), termLength);
        }

        m_logMetaDataBuffer.wrap(
            basePtr + (logLength - LogBufferDescriptor::LOG_META_DATA_LENGTH),
            LogBufferDescriptor::LOG_META_DATA_LENGTH);
    }
    else
    {
        const std::int64_t metaDataSectionOffset = (index_t) (termLength * LogBufferDescriptor::PARTITION_COUNT);
        const std::int64_t metaDataSectionLength = (index_t) (logLength - metaDataSectionOffset);

        m_memoryMappedFiles.push_back(
            MemoryMappedFile::createNew(location, metaDataSectionOffset, metaDataSectionLength));

        std::uint8_t *metaDataBasePtr = m_memoryMappedFiles[0]->getMemoryPtr();

        for (int i = 0; i < LogBufferDescriptor::PARTITION_COUNT; i++)
        {
            // one map for each term
            m_memoryMappedFiles.push_back(MemoryMappedFile::mapExisting(location, i * termLength, termLength));

            std::uint8_t *basePtr = m_memoryMappedFiles[i + 1]->getMemoryPtr();

            if (!useSparseFiles)
            {
                allocatePages(basePtr, m_memoryMappedFiles[i + 1]->getMemorySize());
            }

            m_buffers[i].wrap(basePtr, termLength);
        }

        m_logMetaDataBuffer.wrap(metaDataBasePtr, LogBufferDescriptor::LOG_META_DATA_LENGTH);
    }
}

MappedRawLog::~MappedRawLog()
{
    ::unlink(m_location);
}

std::int32_t MappedRawLog::termLength()
{
    return m_termLength;
}

const char* MappedRawLog::logFileName()
{
    return m_location;
}

void MappedRawLog::allocatePages(std::uint8_t *mapping, size_t length)
{
    size_t pageLength = MemoryMappedFile::getPageSize();

    for (int i = 0; i < length; i += pageLength)
    {
        mapping[i] = 0;
    }
}

}}}